Gems/Atom/Utils/Code/Source/ImGuiGpuProfiler.cpp

```diff
@@ -1408,6 +1441,113 @@ namespace AZ
             ImGui::End();
         }
 
+        void ImGuiGpuMemoryView::UpdateTreemaps()
+        {
+            if (!m_hostTreemap)
+            {
+                if (auto treemapFactory = Profiler::ImGuiTreemapFactory::Interface::Get())
+                {
+                    m_hostTreemap = &treemapFactory->Create(AZ::Name{ "Atom Host Memory Treemap" }, "MiB");
+                    m_hostTreemap->AddMask("Hide Unused", 0);
+                    m_deviceTreemap = &treemapFactory->Create(AZ::Name{ "Atom Device Memory Treemap" }, "MiB");
+                    m_deviceTreemap->AddMask("Hide Unused", 0);
+                }
+            }
+
+            if (m_hostTreemap)
+            {
+                using Profiler::TreemapNode;
+                AZStd::vector<TreemapNode> hostNodes;
+                AZStd::vector<TreemapNode> deviceNodes;
+                for (auto& pool : m_savedPools)
+                {
+                    size_t hostBytes = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Host).m_residentInBytes;
+                    size_t deviceBytes = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Device).m_residentInBytes;
+                    if (hostBytes > 0)
+                    {
+                        hostNodes.push_back();
+                        hostNodes.back().m_name = pool.m_name;
+                    }
+                    else if (deviceBytes > 0)
+                    {
+                        deviceNodes.push_back();
+                        deviceNodes.back().m_name = pool.m_name;
+                    }
+                }
+
+                for (auto& pool : m_savedPools)
+                {
+                    size_t hostBytes = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Host).m_reservedInBytes;
+                    size_t hostResidentBytes = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Host).m_residentInBytes;
+                    size_t deviceBytes = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Device).m_reservedInBytes;
+                    size_t deviceResidentBytes = pool.m_memoryUsage.GetHeapMemoryUsage(RHI::HeapMemoryLevel::Device).m_residentInBytes;
+
+                    TreemapNode* poolNode = nullptr;
+
+                    if (hostBytes > 0)
+                    {
+                        hostNodes.push_back();
+                        poolNode = &hostNodes.back();
+                        poolNode->m_name = pool.m_name;
+                    }
+                    else if (deviceBytes > 0)
+                    {
+                        deviceNodes.push_back();
+                        poolNode = &deviceNodes.back();
+                        poolNode->m_name = pool.m_name;
+                    }
```

Comment:
Slight deja-vu here, a comment explaining what each loop is for and how the second differs from the first would help people reading the code.

Reply:
Yea what this is doing is sending data from each pool to either the device (vram) treemap or the host (system ram) treemap. This code is a little ugly because the original data is structured in a way that pools aren't just tagged with where there memory exists. I plan on refactoring that in the future, which would make the ingestion of the data cleaner. For now I can leave a comment as you suggest.

Reply:
nit: +1 on that - a short sentence before each loop will improve developers understanding and reduce maintenance time.

Reply:
Actually @antonmic now that I'm more awake I'm realizing that the first loop there is supposed to be deleted but I forgot to add my changes to this file to the commit. This code as written actually works because the treemap internally validates and prunes the empty nodes. I'll fix the extraneous code and add a comment.